watertap/core/control_volume_isothermal.py

```diff
@@ -0,0 +1,95 @@
+from idaes.core.base.control_volume_base import ControlVolumeBlockData as _CV
+from idaes.core.base.control_volume0d import ControlVolume0DBlockData as _CV0D
+from idaes.core.base.control_volume1d import ControlVolume1DBlockData as _CV1D
+from idaes.core.util.exceptions import ConfigurationError
+
+import idaes.core.util.scaling as iscale
+
+
+class _IsothermalEnergyBalanceCheckerMixin:
+    def _validate_isothermal_energy_balance_compatibility(self):
+        if self._constructed_energy_balance_type is not EnergyBalanceType.none:
+            raise ConfigurationError(
+                "When using the isothermal assumption, the EnergyBalanceType must be EnergyBalanceType.none"
+            )
+
+    def add_energy_balances(self, balance_type=EnergyBalanceType.useDefault, **kwargs):
+        super().add_energy_balances(balance_type=balance_type, **kwargs)
+        if hasattr(self, "isothermal_assumption_eq"):
+            self._validate_isothermal_energy_balance_compatibility()
+
+    def add_isothermal_assumption(self):
+        if hasattr(self, "_constructed_energy_balance_type"):
+            self._validate_isothermal_energy_balance_compatibility()
+
+
+# put the usual docstring on add_energy_balances
+_IsothermalEnergyBalanceCheckerMixin.add_energy_balances.__doc__ = (
+    _CV.add_energy_balances.__doc__
+)
+
+
+@declare_process_block_class("ControlVolume0DBlock")
+class ControlVolume0DBlockData(_IsothermalEnergyBalanceCheckerMixin, _CV0D):
+    def add_isothermal_assumption(self):
+        super().add_isothermal_assumption()
+
+        @self.Constraint(
+            self.flowsheet().time,
+            doc="Isothermal assumption for 0D control volume",
+        )
+        def isothermal_assumption_eq(b, t):
+            return b.properties_out[t].temperature == b.properties_in[t].temperature
+
+    def calculate_scaling_factors(self):
+        super().calculate_scaling_factors()
+        if hasattr(self, "isothermal_assumption_eq"):
+            for t, c in self.isothermal_assumption_eq.items():
+                sf = iscale.get_scaling_factor(
+                    self.properties_in[t].temperature, default=1e-2, warning=True
+                )
+                iscale.constraint_scaling_transform(c, sf, overwrite=False)
+
+
+@declare_process_block_class("ControlVolume1DBlock")
+class ControlVolume1DBlockData(_IsothermalEnergyBalanceCheckerMixin, _CV1D):
+    def add_isothermal_assumption(self):
+        super().add_isothermal_assumption()
+
+        @self.Constraint(
+            self.flowsheet().time,
+            self.length_domain,
+            doc="Isothermal assumption for 1D control volume",
+        )
+        def isothermal_assumption_eq(b, t, x):
+            if x == b.length_domain.first():
+                return Constraint.Skip
+            return (
+                b.properties[t, b.length_domain.first()].temperature
+                == b.properties[t, x].temperature
+            )
```

Comment:
I know this would be a separate PR, but creating this WaterTAP control volume has me thinking... What if we make a WaterTAP control volume that has a configuration argument for whether it is 0D or 1D and it automatically includes the references for the 0D so that it quacks like a 1D model with `properties[t, x]` instead of just `properties_in[t]` and `properties_out[t]`

Reply:
My only caution here is to think through what this means downstream and what confusions this might create. IDAES deliberately kept the two separate and with different APIs so that it was clear to the end-user which was which, even if it meant they had to understand a bit more.

Reply:
I think there's a trade-off here between keeping separate, understandable APIs while also allowing code-sharing between 0D unit models and their 1D equivalent. I would also add that our 0D-model still allow you to access the 0D-type API.

I would be in favor of enabling a 1D-quacking 0DCV, with the caveat that for models for which we only have a 0D version of we stick to the 0D API in the implementation.

Reply:
Question: is "1D-quacking 0DCV" referring to a 1D-cv with 1 finite element?

Reply:
No -- it has (for a flow-forward model) `properties_in[t]` mapped to `properties[0,t]` and `properties_out[t]` mapped to `properties[1,t]` using references.

This allows for the expression of constraints common to both a 0D and 1D version of a model, see, e.g., https://github.com/watertap-org/watertap/blob/d2382933c88859b6e4f6ae7f791ca187f30a3a33/watertap/unit_models/reverse_osmosis_base.py#L346-L372